x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+                    this.allowedAudiences
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug("Realm [%s] allows audiences %s. Allowed audiences are %s.", super.name(), jwtAudiences, this.allowedAudiences);
+
+            // TODO The implementation of JWT authentication will be completed in a later PR
+            // At this point, this is the right realm to do validation. Trigger AuthenticationResult.unsuccessful() if any problems found.
+
+            // Cache lookup
+            // 1. If present in cache, AuthenticationResult.successful().
+
+            // JWT Authentication
+            // 1. Verify signature (HMAC or RSA)
+            // 2. If present verify nfb <= iat.
+            // 3. If present verify nfb <= exp.
+            // 4. If present verify iat <= exp.
+            // 5. If present, verify nfb + allowedClockSkew < now.
+            // 6. If present, verify iat + allowedClockSkew >= now.
+            // 7. If present, verify exp + allowedClockSkew >= now.
+
+            // Client Authorization
+            switch (this.clientAuthorizationType) {
+                case JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_SHARED_SECRET:
+                    if (this.clientAuthorizationSharedSecret.equals(clientAuthorizationSharedSecret) == false) {
+                        final String msg = String.format(
+                            Locale.ROOT,
+                            "Realm [%s] client authentication failed for [%s].",
+                            Locale.ROOT,
+                            super.name(),
+                            this.clientAuthorizationType
+                        );
+                        LOGGER.debug(msg);
+                        listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                        return;
+                    }
+                    LOGGER.debug("Realm [%s] client authentication succeeded for [%s].", super.name(), this.clientAuthorizationType);
+                    break;
+                case JwtRealmSettings.SUPPORTED_CLIENT_AUTHORIZATION_TYPE_NONE:
+                default:
+                    LOGGER.debug("Realm [%s] client authentication skipped for [%s].", super.name(), this.clientAuthorizationType);
+                    break;
+            }
+
+            final String jwtPrincipal = this.principalAttribute.getClaimValue(jwtClaimsSet);
+            final String msg1 = String.format(
+                Locale.ROOT,
+                "Realm [%s] got principal [%s] from claim [%s] and parser [%s]. JWTClaimsSet is [%s].",
+                super.name(),
+                jwtPrincipal,
+                this.principalAttribute.getName(),
+                this.principalAttribute.getParser().toString(),
+                jwtClaimsSet.toString()
+            );
+            LOGGER.debug(msg1);
```

Comment:
```suggestion
            LOGGER.trace(
                "Realm [{}] got principal [{}] from claim [{}] and parser [{}]. JWTClaimsSet is [{}].",
                super.name(),
                jwtPrincipal,
                this.principalAttribute.getName(),
                this.principalAttribute.getParser(),
                jwtClaimsSet
            );
```

Reply:
I prefer debug logs to show the customer why some JWT realms are skipped before the authenticating JWT realm is found to attempt JWT validation.

Reply:
Removing Locale.ROOT causes this build check fail, so I can't remove that parameter.
`./gradlew loggerUsageCheck`

Reply:
Done

Reply:
Discussed two days ago. I think we agreed it was OK.